physics/kepler_orbit_body.hpp

```diff
@@ -157,94 +213,478 @@ KeplerOrbit<Frame>::KeplerOrbit(
+  auto& period = elements.period;
+  auto& hyperbolic_mean_motion = elements.hyperbolic_mean_motion;
+  auto& hyperbolic_excess_velocity = elements.hyperbolic_excess_velocity;
+  auto& semiminor_axis = elements.semiminor_axis;
+  auto& impact_parameter = elements.impact_parameter;
+  auto& semilatus_rectum = elements.semilatus_rectum;
+  auto& specific_angular_momentum = elements.specific_angular_momentum;
+  auto& periapsis_distance = elements.periapsis_distance;
+  auto& apoapsis_distance = elements.apoapsis_distance;
+
+  CompleteConicParametersByCategory(elements, μ);
+
+  CHECK_EQ(eccentricity.has_value() + semimajor_axis.has_value() +
+               semiminor_axis.has_value() + semilatus_rectum.has_value() +
+               periapsis_distance.has_value() + apoapsis_distance.has_value(),
+           2) << elements;
+
+  // TODO(egg): some of these formulae are very ill-conditioned near the
+  // parabolic case, and can be easily rewritten.  Investigate.
+  if (eccentricity && semimajor_axis) {
+    double const& e = *eccentricity;
+    Length const& a = *semimajor_axis;
+    semiminor_axis = a * Sqrt(1 - Pow<2>(e));
+    impact_parameter = -a * Sqrt(Pow<2>(e) - 1);
+    semilatus_rectum = a * (1 - Pow<2>(e));
+    periapsis_distance = a * (1 - e);
+    apoapsis_distance = a * (1 + e);
+  } else if (eccentricity && semiminor_axis) {
+    double const& e = *eccentricity;
+    Length const& abs_b = e > 1 ? *impact_parameter : *semiminor_axis;
+    semilatus_rectum = abs_b * Sqrt(Abs(1 - Pow<2>(e)));
+    semimajor_axis = *semilatus_rectum / (1 - Pow<2>(e));
+    periapsis_distance = *semimajor_axis * (1 - e);
+    apoapsis_distance = *semimajor_axis * (1 + e);
+  } else if (eccentricity && semilatus_rectum) {
+    double const& e = *eccentricity;
+    Length const& ℓ = *semilatus_rectum;
+    semimajor_axis = ℓ / (1 - Pow<2>(e));
+    semiminor_axis = *semimajor_axis * Sqrt(1 - Pow<2>(e));
+    impact_parameter = -*semimajor_axis * Sqrt(Pow<2>(e) - 1);
+    periapsis_distance = ℓ / (1 + e);
+    apoapsis_distance = ℓ / (1 - e);
+  } else if (eccentricity && periapsis_distance) {
+    double const& e = *eccentricity;
+    Length const& r_pe = *periapsis_distance;
+    semimajor_axis = r_pe / (1 - e);
+    semiminor_axis = *semimajor_axis * Sqrt(1 - Pow<2>(e));
+    impact_parameter = -*semimajor_axis * Sqrt(Pow<2>(e) - 1);
+    semilatus_rectum = r_pe * (1 + e);
+    apoapsis_distance = *semimajor_axis * (1 + e);
+  } else if (eccentricity && apoapsis_distance) {
+    double const& e = *eccentricity;
+    Length const& r_ap = *apoapsis_distance;
+    semimajor_axis = r_ap / (1 + e);
+    semiminor_axis = *semimajor_axis * Sqrt(1 - Pow<2>(e));
+    impact_parameter = -*semimajor_axis * Sqrt(Pow<2>(e) - 1);
+    semilatus_rectum = r_ap * (1 - e);
+    periapsis_distance = *semimajor_axis * (1 - e);
+  } else if (semimajor_axis && semiminor_axis) {
+    Length const& a = *semimajor_axis;
+    auto const& b² = *semiminor_axis != *semiminor_axis
```

Comment:
Should not be a reference if it comes from a computation.  Same comment elsewhere in this function.